### PR TITLE
Resolve the problem of the generating of BroadcastJoin in physical plan

### PR DESCRIPTION
**Pull request by Guangggggg**
### What problem does this PR solve? 

A small table can not be generated BroadcastJoin Physical Plan.

The table size is less than default value of 'spark.sql.autoBroadcastJoinThreshold' .

And although TableSizeEstimator.estimatedTableSize < 'spark.sql.autoBroadcastJoinThreshold' , the physical plan is not still BroadcastJoin.

In some cases, this will reduce the execution efficiency of tispark, for example, when the user sets 'spark.sql.adaptive.enabled=false', the table will never be brocast

### What is changed and how it works?

This PR create the 'TiDBTableScan' class to provide 'sizeInBytes' of table when generate physical plan, which can be used to determined whether to broadcast a table. And Scan class is common in other file formats such as parquet，orc and so on.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

```
//create a table in tidb
CREATE TABLE person (id INT(11),name VARCHAR(255),birthday DATE);
INSERT INTO person VALUES(1,'tom','20170912');
INSERT INTO person VALUES(2,'tom','20170912');
INSERT INTO person VALUES(3,'tom','20170912');
INSERT INTO person VALUES(4,'tom','20170912');
INSERT INTO person VALUES(5,'tom','20170912');
INSERT INTO person VALUES(6,'tom','20170912');
INSERT INTO person VALUES(7,'tom','20170912');
INSERT INTO person VALUES(8,'tom','20170912');
INSERT INTO person VALUES(9,'tom','20170912');
```
```
SparkSession spark = SparkSession.builder().master("local[*]")
            .config("spark.sql.extensions","org.apache.spark.sql.TiExtensions")
            .config("spark.sql.catalog.tidb_catalog","org.apache.spark.sql.catalyst.catalog.TiCatalog")
            .config("spark.tispark.pd.addresses","127.0.0.1:2379")
            .config("spark.sql.catalog.tidb_catalog.pd.addresses","127.0.0.1:2379")
            .getOrCreate();
Dataset<Row> dataset = spark.sql("select * from tidb_catalog.test.person t1 left join tidb_catalog.test.person t2 on t1.id = t2.id");
     dataset.explain(true);
```
```
===========================result:physical plan before modification==============================

*(5) SortMergeJoin [id#6L], [id#9L], LeftOuter
:- *(2) Sort [id#6L ASC NULLS FIRST], false, 0
:     +- Exchange hashpartitioning(id#6L, 200), ENSURE_REQUIREMENTS, [id=#27]
:       +- *(1) ColumnarToRow
:         +- TiKV CoprocessorRDD{[table: person] TableReader, Columns: id@LONG, name@VARCHAR(255)....
+- *(4) Sort [id#9L ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(id#9L, 200), ENSURE_REQUIREMENTS, [id=#35]
        +- *(3) ColumnarToRow
          +- TiKV CoprocessorRDD{[table: person] TableReader, Columns: id@LONG, name@VARCHAR(255)....

===========================result:physical plan after modification==============================

*(2) BroadcastHashJoin [id#6L], [id#9L], LeftOuter, BuildRight, false
:- *(2) ColumnarToRow
:    +- TiKV CoprocessorRDD{[table: person] TableReader, Columns: id@LONG, name@VARCHAR(255).....
+- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, true]),false), [id=#22]
     +- *(1) ColumnarToRow
        +- TiKV CoprocessorRDD{[table: person] TableReader, Columns: id@LONG, name@VARCHAR(255)...
```
Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScan.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  override def readSchema(): StructType = schema
+
+  override def estimateStatistics(): Statistics = {
+    new Statistics {
+      override def sizeInBytes(): OptionalLong = {
+        val size = tableRef.sizeInBytes
+        OptionalLong.of(size)
+      }
+      override def numRows(): OptionalLong = OptionalLong.empty()
```
Comment (shiyuhang0) [inline core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScan.scala:23]: Can you explain why numRows can be set as empty? Will it influence the execution plan?

if we need numRows, we can  add `numRows` in tableRef, and set the value of `numRows` in `loadStatistics` (where the sizeInBytes will be set)

**Comment by shiyuhang0** [comment]
thank you for your contribute Please
1. fix the fmt by  `mvn mvn-scalafmt_2.12:format -Dscalafmt.skip=false`
2. add license for the new added file. you can copy one from other files. remember to change the `year` to 2013

File: core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScanBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.pingcap.tispark.v2
+
+import com.pingcap.tispark.TiTableReference
+import org.apache.spark.sql.connector.read.{Scan, ScanBuilder}
+import org.apache.spark.sql.types.StructType
+
+class TiDBTableScanBuilder(
```
Comment (shiyuhang0) [inline core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScanBuilder.scala:7]: use case class. `JDBCScanBuilder` is also case class

File: core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScan.scala (hunk 1 above)
Comment (Guangggggg) [reply core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScan.scala:37]: 1.Can you explain why numRows can be set as empty?
Like the FileScan.class in the Spark source code, `numRows` has not been calculated. It's not necessary.

2.Will it influence the execution plan?
`sizeInBytes` can meet the most requirements in generating physical plans, but `numRows` will makes the size estimation result more accurate.

The process of size estimation can be found in SizeInBytesOnlyStatsPlanVisitor.class in the Spark source code

File: core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScanBuilder.scala (hunk 2 above)
Comment (Guangggggg) [reply core/src/main/scala/com/pingcap/tispark/v2/TiDBTableScanBuilder.scala]: It has been modified. thx.

**Review by shiyuhang0 — APPROVED** [review]

**Comment by shiyuhang0** [comment]
LGTM~

**Review by shiyuhang0 — APPROVED** [review]

**Comment by shiyuhang0** [comment]
@zhangyangyu @xuanyu66 PTAL

**Review by xuanyu66 — APPROVED** [review]

**Pull request merged**